maintenance: add new `cache-local-objects` maintenance task

mjcheetham opened the pull request:
Introduce a new maintenance task, `cache-local-objects`, that operates on Scalar or VFS for Git repositories with a per-volume, shared object cache (specified by `gvfs.sharedCache`) to migrate packfiles and loose objects from the repository object directory to the shared cache.

Older versions of `microsoft/git` incorrectly placed packfiles in the repository object directory instead of the shared cache; this task will help clean up existing clones impacted by that issue.

Fixes #716

mjcheetham committed:
maintenance: add cache-local-objects maintenance task

Introduce a new maintenance task, `cache-local-objects`, that operates
on Scalar or VFS for Git repositories with a per-volume, shared object
cache (specified by `gvfs.sharedCache`) to migrate packfiles and loose
objects from the repository object directory to the shared cache.

Older versions of `microsoft/git` incorrectly placed packfiles in the
repository object directory instead of the shared cache; this task will
help clean up existing clones impacted by that issue.

Migration of packfiles involves the following steps for each pack:

1. Hardlink (or copy):
   a. the .pack file
   b. the .keep file
   c. the .rev file
2. Move (or copy + delete) the .idx file
3. Delete/unlink:
   a. the .pack file
   b. the .keep file
   c. the .rev file

Moving the index file after the others ensures the pack is not read
from the new cache directory until all associated files (rev, keep)
exist in the cache directory also.

Moving loose objects operates as a move, or copy + delete.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

dscho commented on Documentation/git-maintenance.txt:
```diff
@@ -158,6 +159,13 @@ pack-refs::
 	for more information.
 
+vfs-cache-move::
+	The `vfs-cache-move` task operates only Scalar or VFS for Git
```
I think the word "on" is missing between "only" and "Scalar", not that that matters a lot...

dscho commented on builtin/gc.c:
```diff
@@ -1345,6 +1345,154 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
 }
 
+
+static void copy_file(const char *srcdir, const char *dstdir, const char *name)
```
There's already `copy_file()` in `copy.h`, can we use that instead?

mjcheetham replied:
Done!

dscho commented on builtin/gc.c:
```diff
@@ -1345,6 +1345,154 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+	basename = xstrndup(pack_filename, strlen(pack_filename) - 5 /*.pack*/);
+	keep_filename = xstrfmt("%s.keep", basename);
+	rev_filename = xstrfmt("%s.rev", basename);
+	idx_filename = xstrfmt("%s.idx", basename);
```
What should we do with `.pack` files that have no associated `.idx` file?

(Could be an in-progress `git fetch`? Although that would probably have a `tmp_*` file at first, there is a small time window when it is already renamed to `.pack` but the `.idx` has not been renamed yet; But it could also be a left-over from some failed operation...).

mjcheetham replied:
We now skip if the `.idx` file is missing, meaning we need both the `.idx` and the `.pack` to start migration. If we just miss the `.idx` creation, then we'll catch it next time the maintenance is run.

dscho commented on builtin/gc.c:
```diff
@@ -1345,6 +1345,154 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+	strbuf_release(&path);
+
+	/* Copy all but the index file, which we will *move* atomically */
+	copy_file(srcdir, dstdir, pack_filename);
```
I would like to try hard-linking it first, i.e. the following strategy:

1. create a hard-link of the `.pack` file to the new location
2. create hard-links of the corresponding `.keep`, `.rev` and finally `.idx` files, too
3. remove the original `.idx` file, then the `.pack`, `.rev` and `.keep` file, too

Of course, hard-linking can fail, e.g. when the object directories are on different drives. In that instance, we do have to fall back to copying instead (but keeping the operations in the same order).

dscho commented on t/t7900-maintenance.sh:
```diff
@@ -1011,4 +1011,34 @@ test_expect_success 'repacking loose objects is quiet' '
+		touch .git/objects/pack/vfs-12345678.pack &&
+		touch .git/objects/pack/vfs-12345678.keep &&
+		touch .git/objects/pack/vfs-12345678.rev &&
+		touch .git/objects/pack/vfs-12345678.idx &&
```
I fear that we have to make this a proper `.pack` and `.idx`; This is relatively easy, though:

```sh
printf 'blob\ndata <<END\n%s\nEND\n\n' 1 2 3 4 5 |
git -c fastimport.unpackLimit=0 fast-import
```

The only non-obvious thing is to find out what the pack file is called, as `fast-import` does not report _that_ ;-)

derrickstolee commented on builtin/gc.c:
```diff
@@ -1345,6 +1348,131 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+	free(pack_src);
+	free(pack_dst);
+	free(basename);
+
```
One way to avoid some of this repeated logic would be to have an array of pairs (from/src) where if the first one is non-NULL then we act upon it. The order of the array can be constructed to follow the order we want. It can then be expanded to include a new file type (e.g. `.promisor`, `.mtimes` though they shouldn't happen here) with minimal code duplication.

derrickstolee commented on builtin/gc.c:
```diff
@@ -1504,6 +1637,8 @@ static void initialize_maintenance_strategy(void)
 		tasks[TASK_LOOSE_OBJECTS].schedule = SCHEDULE_DAILY;
 		tasks[TASK_PACK_REFS].enabled = 1;
 		tasks[TASK_PACK_REFS].schedule = SCHEDULE_WEEKLY;
+		tasks[TASK_VFS_CACHE_MOVE].enabled = 1;
```
This is modifying the `incremental` strategy for everyone, but this task is a no-op. That should be fine.

In older versions, `scalar` would configure the schedule manually, but [that appears to no longer be the case](https://github.com/microsoft/git/blob/1d7817ccfe14b7c0b9e2f9831cc1b6f0214b663f/scalar.c#L161) and instead we rely on [the default behavior of `git maintenance start`](https://github.com/microsoft/git/blob/1d7817ccfe14b7c0b9e2f9831cc1b6f0214b663f/scalar.c#L254-L265). So this is the cleanest way to upgrade folks with Scalar enlistments.

You may also want to add this task [to `scalar run`](https://github.com/microsoft/git/blob/1d7817ccfe14b7c0b9e2f9831cc1b6f0214b663f/scalar.c#L1160-L1166).

mjcheetham replied:
> You may also want to add this task [to scalar run](https://github.com/microsoft/git/blob/1d7817ccfe14b7c0b9e2f9831cc1b6f0214b663f/scalar.c#L1160-L1166).

Not sure if this is required for newer tasks. The task mapping in: https://github.com/microsoft/git/blob/1d7817ccfe14b7c0b9e2f9831cc1b6f0214b663f/scalar.c#L1160-L1166
..feels primarily for backwards compatibility, for users/scripts using the `scalar` command to run those tasks that existed before with different names, rather than their `git maintenance`. names/command.

Thoughts @dscho?

derrickstolee replied:
Sometimes I have recommended users run `scalar run` instead of `git maintenance run --schedule=weekly` for simplicity. But I could change that habit.

dscho replied:
I do like the option of asking users to call `scalar run` to clean up things like the `.pack` files that should go to the shared object directory.

mjcheetham replied:
Fair point on the ease of use! Added.

mjcheetham committed:
fixup! maintenance: care about gvfs.sharedCache config

mjcheetham committed:
t7900-maintenance.sh: reset config between tests

Tests in t7900 assume the state of the `maintenance.strategy`
config setting; set/unset by previous tests. Correct this by
explictly unsetting and re-setting the config at the start of the
tests.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
scalar.c: add cache-local-objects task

Add the `cache-local-objects` maintenance task to the list of tasks run
by the `scalar run` command. It's often easier for users to run the
shorter `scalar run` command than the equivalent `git maintenance`
command.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham commented:
@derrickstolee @dscho I've addressed the issues raised in the comments so far. Please could I have another look over? Thanks! :)

dscho commented on builtin/gc.c:
```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+	if (!link(src, dst))
+		return;
+
+	warning_errno(_("failed to link '%s' to '%s'"), src, dst);
```
Do we really want to warn always? I just verified that `link()` fails with `errno == EXDEV` when trying to link between different drives, maybe we want to suppress that warning under this condition?

dscho commented on builtin/gc.c:
```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+
+	warning_errno(_("failed to link '%s' to '%s'"), src, dst);
+
+	if (!copy_file(dst, src, 0644))
```
Not that it matters a lot on Windows, but I think `0444` would be more appropriate here; All the files in my object directory on Linux are marked as `-r--r--r--`.

dscho commented on builtin/gc.c:
```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+		strbuf_addf(&dst, "%s/%s%s", dstdir, basename, files[i].ext);
+
+		if (files[i].move) {
+			if (rename(src.buf, dst.buf))
```
I wonder whether `rename()` in a cross-drive scenario would result in a partial `.idx` file on the target side. Do you know @mjcheetham?

mjcheetham replied:
`rename` would not work in cross-drive scenarios at all - good point. Updated to fall-back to using `copy_file && unlink` in this scenario.

dscho commented on builtin/gc.c:
```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+
+	srcdir = xstrndup(full_path, full_path_len - strlen(file_name) - 1);
+
+	/* No cache or same source + desintation means there's no work to do. */
```
`s/desintation/destination/` 😊

dscho commented on builtin/gc.c:
```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+	srcdir = xstrndup(full_path, full_path_len - strlen(file_name) - 1);
+
+	/* No cache or same source + desintation means there's no work to do. */
+	if (!vfs_object_dir || !fspathcmp(srcdir, vfs_object_dir))
```
It would probably make sense to move the early return upon `!vfs_object_dir` to the `maintenance_task_vfs_cache_move()` function, otherwise `git maintenance` would do a lot of unnecessary work.

dscho commented on builtin/gc.c:
```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+
+	trace2_region_enter("maintenance", "migrate_pack", the_repository);
+
+	basename = xstrndup(pack_filename, strlen(pack_filename) - 5 /*.pack*/);
```
An alternative that would (at least to me) look more consistent with Git's opinionated source code style would be to set `size_t len = strlen(pack_filename) - 5; /* .pack */` and then using `strbuf_addf(&src, "%s/%.*s.idx", srcdir, len, pack_filename);` below, i.e. avoid copying the basename explicitly.

An even more consistent way would be to populate `src` and `dst` with the full path to the pack file, then use `strbuf_strip_suffix()` on it and then store the lengths and reset them via `strbuf_setlen()` instead of `strbuf_reset()`. Not sure whether you want to do that though?

mjcheetham replied:
`"%s/%.*s.idx"` cannot be used when `len` is a `size_t`, unless we cast to `int`?

dscho replied:
Right. Better use the `strbuf_strip_suffix()` route, then?

mjcheetham replied:
I'll just move to `strbuf_add`

dscho replied:
If you like the end result better, it is certainly fine by me.

dscho commented on builtin/gc.c:
```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+
+	/* A pack without an index file is not yet ready to be migrated. */
+	strbuf_addf(&src, "%s/%s%s", srcdir, basename, ".idx");
+	if (stat(src.buf, &st))
```
It would probably be more idiomatic to write `file_exists(src.buf)` here, but then, it would not buy us much because that function does the exact same thing (without bothering to verify that the entity is actually a _file_). So: 🤷

dscho commented on builtin/gc.c:
```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+	if (!vfs_object_dir || !fspathcmp(srcdir, vfs_object_dir))
+		goto cleanup;
+
+	strbuf_addf(&dstdir, "%s/pack", vfs_object_dir);
```
Would probably make sense to initialize this in `maintenance_task_vfs_cache_move()` and pass it in via `data`?

dscho commented on builtin/gc.c:
```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+	trace2_region_leave("maintenance", "migrate_pack", the_repository);
+}
+
+static void move_pack_to_vfs_cache(const char *full_path, size_t full_path_len,
+				   const char *file_name, UNUSED void *data)
```
It might make sense to merge `migrate_pack()` directly into this function, after the `.pack` suffix check.

dscho commented on t/t7900-maintenance.sh:
```diff
@@ -1012,4 +1012,129 @@ test_expect_success 'repacking loose objects is quiet' '
 '
 
+test_expect_success 'vfs-cache-move task with no shared cache no op' '
+	test_when_finished "rm -rf repo" &&
+	git init repo &&
+	(
+		cd repo &&
+
+		test_commit something &&
+		git config set maintenance.gc.enabled false &&
+		git config set maintenance.vfs-cache-move.enabled true &&
+		git config set maintenance.vfs-cache-move.auto 1 &&
+
+		printf "blob\ndata <<END\n%s\nEND\n\n" 1 2 3 4 5 | \
+		git -c fastimport.unpackLimit=0 fast-import &&
+		find .git/objects/pack \
+			-type f \
+			\( -name "*.pack" \
+			-o -name "*.idx" \
+			-o -name "*.keep" \
+			-o -name "*.rev" \) >files.txt &&
```
This seems to be repeated code. Could we please move it into a shell function that is then used in those four new test cases?

mjcheetham replied:
Done

dscho commented on t/t7900-maintenance.sh:
```diff
@@ -1012,4 +1012,129 @@ test_expect_success 'repacking loose objects is quiet' '
+
+		rm -f .git/objects/pack/*.rev .git/objects/pack/*.keep &&
+
+		sed "s|.*/|../cache/pack/|" src.txt >dst.txt &&
```
Isn't `s|...|...|` a GNU extension? I.e. does this work on macOS? From https://pubs.opengroup.org/onlinepubs/9699919799/utilities/sed.html it looks to me as if only `s/.../.../` would be portable...

derrickstolee commented on builtin/gc.c:
```diff
@@ -1345,6 +1350,155 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
+
+		strbuf_setlen(&src, srclen);
+		strbuf_addstr(&src, files[i].ext);
+		if (unlink(src.buf))
+			warning_errno(_("failed to delete '%s'"), src.buf);
```
This may need some updating, as I got the following error when testing locally:

```
warning: failed to delete '.git/objects/pack/packed-manually-3d4517d7d49dc7f09b64773b940186047f4dac16.keep': No such file or directory
```

So should we check for existence first? Or does `unlink()` return a value indicating that the file doesn't exist so we can ignore the error?

mjcheetham replied:
Ah. Good find. I'm missing a `file_exists(src.buf)` check, like in the first loop above. Adding...

mjcheetham replied:
Added now in a `fixup!`

dscho reviewed (dismissed):
Looks good to me!

derrickstolee reviewed (approved):
Thanks, @mjcheetham! I tested this with my local Office enlistment and it worked great to move both loose objects and packfiles. I'm excited for this to help our customers!
